lib/src/request.dart

```diff
@@ -3,159 +3,94 @@
 // BSD-style license that can be found in the LICENSE file.
 
 import 'dart:convert';
-import 'dart:typed_data';
 
-import 'package:http_parser/http_parser.dart';
-
-import 'base_request.dart';
-import 'byte_stream.dart';
+import 'message.dart';
 import 'utils.dart';
 
-/// An HTTP request where the entire request body is known in advance.
-class Request extends BaseRequest {
-  /// The size of the request body, in bytes. This is calculated from
-  /// [bodyBytes].
-  ///
-  /// The content length cannot be set for [Request], since it's automatically
-  /// calculated from [bodyBytes].
-  int get contentLength => bodyBytes.length;
-
-  set contentLength(int value) {
-    throw new UnsupportedError("Cannot set the contentLength property of "
-        "non-streaming Request objects.");
-  }
-
-  /// The default encoding to use when converting between [bodyBytes] and
-  /// [body]. This is only used if [encoding] hasn't been manually set and if
-  /// the content-type header has no encoding information.
-  Encoding _defaultEncoding;
-
-  /// The encoding used for the request. This encoding is used when converting
-  /// between [bodyBytes] and [body].
-  ///
-  /// If the request has a `Content-Type` header and that header has a `charset`
-  /// parameter, that parameter's value is used as the encoding. Otherwise, if
-  /// [encoding] has been set manually, that encoding is used. If that hasn't
-  /// been set either, this defaults to [UTF8].
-  ///
-  /// If the `charset` parameter's value is not a known [Encoding], reading this
-  /// will throw a [FormatException].
-  ///
-  /// If the request has a `Content-Type` header, setting this will set the
-  /// charset parameter on that header.
-  Encoding get encoding {
-    if (_contentType == null ||
-        !_contentType.parameters.containsKey('charset')) {
-      return _defaultEncoding;
-    }
-    return requiredEncodingForCharset(_contentType.parameters['charset']);
-  }
-
-  set encoding(Encoding value) {
-    _checkFinalized();
-    _defaultEncoding = value;
-    var contentType = _contentType;
-    if (contentType == null) return;
-    _contentType = contentType.change(parameters: {'charset': value.name});
-  }
-
-  // TODO(nweiz): make this return a read-only view
-  /// The bytes comprising the body of the request. This is converted to and
-  /// from [body] using [encoding].
+/// Represents an HTTP request to be processed by a `http` application.
+class Request extends Message {
+  /// The HTTP method of the request. Most commonly "GET" or "POST", less
+  /// commonly "HEAD", "PUT", or "DELETE". Non-standard method names are also
+  /// supported.
+  final String method;
+
+  /// The URL to which the request will be sent.
+  final Uri url;
+
+  /// Creates a new [Request] for [url] using [method].
```

Comment:
Before I start copy pasting it everywhere how does this sound?

```
  /// Creates a new [Request] for [url] using [method].
  ///
  /// [body] is the request body. It may be either a [String], a [List<int>], a
  /// [Stream<List<int>>], or `null` to indicate no body. If it's a [String],
  /// [encoding] is used to encode it to a [Stream<List<int>>]. The default
  /// encoding is UTF-8.
  ///
  /// [headers] are the HTTP headers for the request.
  ///
  /// Extra [context] can be used by middleware and handler to pass information
  /// between inner middleware and handlers.
```